### PR TITLE
First version

### PR DESCRIPTION
**Pull request by axd4**
First version of the digiwf-spring-cloudstream-utils library

Not included in this is the Exception Handling

### CODE REVIEW

File: digiwf-spring-cloudstream-utils/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/PayloadSender.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    public void sendPayload(final Object payload, final String type) {
+        Map<String, Object> headers = new HashMap<>();
+        headers.put(StreamingHeaders.TYPE, type);
+        MessageHeaders messageHeaders = new MessageHeaders(headers);
+        sendPayload(payload, messageHeaders);
+
+    }
+
+    public void sendPayload(final Object payload, final MessageHeaders messageHeaders) {
+
```
Comment (dominikhorn93) [inline digiwf-spring-cloudstream-utils/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/PayloadSender.java:30]: Remove line break? Same style as above

File: digiwf-spring-cloudstream-utils/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/PayloadSender.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        final Message<Object> message = MessageBuilder.createMessage(payload, messageHeaders);
+
+        final Sinks.EmitResult emitResult = this.messageSink.tryEmitNext(message);
```
Comment (dominikhorn93) [inline digiwf-spring-cloudstream-utils/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/PayloadSender.java:33]: We should discuss if we want to go blocking with `tryEmitNext` or if we want to use non-blocking `emitNext` and pass a `EmitFailureHandler` as second parameter. @xdoo what do you think?

File: pom.xml [hunk 3]
```diff
@@ -0,0 +1,93 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (dominikhorn93) [inline pom.xml:1]: we need the same plugins as here: https://github.com/it-at-m/digiwf-s3-integration/blob/dev/pom.xml for releasing it in maven central

**Review by dominikhorn93 — CHANGES_REQUESTED** [review]

**Comment by dominikhorn93** [comment]
Great work! 👍  Have some small change requests.

File: digiwf-spring-cloudstream-utils/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/EventEmitter.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.messaging.Message;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Sinks;
+
+import java.util.function.Supplier;
+
+@Configuration
+public class EventEmitter {
```
[the remark this answers is not in the document]
Comment (axd4) [reply digiwf-spring-cloudstream-utils/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/EventEmitter.java:12]: Can just call it MessageEmitter too. We don't really need Event here imo, my bad :)

File: digiwf-spring-cloudstream-utils/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/PayloadSender.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class PayloadSender {
+
+    private final Sinks.Many<Message<Object>> messageSink;
+
+    public void sendPayload(final Object payload, final String type) {
+        Map<String, Object> headers = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (axd4) [reply digiwf-spring-cloudstream-utils/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/PayloadSender.java]: Done

File: digiwf-spring-cloudstream-utils/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/PayloadSender.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        headers.put(StreamingHeaders.TYPE, type);
+        MessageHeaders messageHeaders = new MessageHeaders(headers);
+        sendPayload(payload, messageHeaders);
+
```
[the remark this answers is not in the document]
Comment (axd4) [reply digiwf-spring-cloudstream-utils/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/PayloadSender.java]: Done

File: README.md [hunk 7]
```diff
@@ -1 +1,269 @@
-# digiwf-spring-cloudstream-utils
\ No newline at end of file
+<div id="top"></div>
```
[the remark this answers is not in the document]
Comment (axd4) [reply README.md:1]: Thanks! ❤️

File: digiwf-spring-cloudstream-utils-starter/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/configuration/StreamingConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+
+    private final StreamingProperties streamingProperties;
+
+    // 1. Reads the message from the specified EventBus topic (see spring.cloud.stream.bindings.functionRouter-in-0.destination)
+    // 2. Routes messages to the corresponding method (based on 'type'-header)
+
+    @Bean
+    public MessageRoutingCallback customRouter() {
+        return new MessageRoutingCallback() {
```
[the remark this answers is not in the document]
Comment (dominikhorn93) [reply digiwf-spring-cloudstream-utils-starter/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/configuration/StreamingConfiguration.java]: Almost 🚀 

Would do it this way:

```
@RequiredArgsConstructor
public class RoutingCallback implements MessageRoutingCallback {

    private final Map<String, String> typeMappings;

    @Override
    public FunctionRoutingResult routingResult(final Message<?> message) {
        final String functionDefinition;

        if (message.getHeaders().containsKey(StreamingHeaders.TYPE)) {
            final String header = (String) message.getHeaders().get(StreamingHeaders.TYPE);
            functionDefinition = this.typeMappings.getOrDefault(header, FUNCTION_ROUTING_ERROR);
        } else {
            functionDefinition = MISSING_TYPE_HEADER_ERROR;
        }

        return new FunctionRoutingResult(functionDefinition);
    }

}
```


```
@Configuration
@RequiredArgsConstructor
@ComponentScan(basePackages = "io.muenchendigital.digiwf.spring.cloudstream.utils")
@EnableConfigurationProperties(StreamingProperties.class)
public class StreamingConfiguration {

    private final StreamingProperties streamingProperties;

    @Bean
    public MessageRoutingCallback customRouter() {
        return new RoutingCallback(this.streamingProperties.getTypeMappings());
    }
}
```

**Comment by dominikhorn93** [comment]
@axd4 We should also add "real" Javadocs. But for that we can also create an issue. What do you think?

File: digiwf-spring-cloudstream-utils/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/PayloadSender.java (hunk 1 above)
Comment (axd4) [reply digiwf-spring-cloudstream-utils/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/PayloadSender.java]: Done

File: pom.xml (hunk 3 above)
Comment (axd4) [reply pom.xml:1]: Implemented, seems to work as expected.

File: digiwf-spring-cloudstream-utils-starter/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/configuration/StreamingConfiguration.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (axd4) [reply digiwf-spring-cloudstream-utils-starter/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/configuration/StreamingConfiguration.java]: Adopted your code

**Comment by axd4** [comment]
> @axd4 We should also add "real" Javadocs. But for that we can also create an issue. What do you think?

Let's do that in a new issue and merge the first version

File: digiwf-spring-cloudstream-utils/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/PayloadSender.java (hunk 2 above)
Comment (axd4) [reply digiwf-spring-cloudstream-utils/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/PayloadSender.java:30]: See #3

**Review by dominikhorn93 — APPROVED** [review]

**Pull request merged**